### PR TITLE
implemented custom ruby enumerable module

### PR DESCRIPTION
**Pull request by chirchir12**
# Custom Enumerable Methods
- the purpose of this project is to implement custom ruby enumerable methods without using built-in ruby methods. in this regard, the following methods were added to myenumerable module file 
1. my_each
2. my_each_with_index
3. my_map
4. my_select
5. my_all
6. my_any
7. my_none
8. my_count
9. my_inject
10. modified my_map to take in Proc instead of block 
11. multiply_els

### CODE REVIEW

**Review by nikasulo — CHANGES_REQUESTED** [review]
# Great work! @chirchir12 
Please ensure you have a descriptive PR description before you request another review. This is a requirement before a review can be done

##### Think I missed something? Need some help with this or any of your projects? Please feel free to reach out to me on [Twitter](https://twitter.com/oluwadamilareo_) or [Instagram](https://instagram.com/oluwadamilare_olusakin), even if you just want to chat and I will respond to you as soon as I can.

**Comment by chirchir12** [comment]
# changes made 
*  The methods implemented include: #my_each, #my_each_with_index, #my_select, # my_all, #my_any ,#my_none ,#my_count, #my_map, #my_inject, my_map and multiply_els.

File: README.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
+# Ruby Enumerable methods
+
+In this project,I implemented a Custom Enumerable method. The methods implemented include: #my_each, #my_each_with_index, #my_select, # my_all, #my_any ,#my_none ,#my_count, #my_map, #my_inject, my_map and multiply_els.
```
Comment (mosesogwo) [inline README.md]: The first sentence will only be correct as:
In this project,I implemented a Custom Enumerable module. or
In this project,I implemented Custom Enumerable methods

File: myenumerables.rb [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  def my_map_with_proc(proc = nil)
+    if block_given? || !proc.nil?
+      new_array = []
+      my_each { |val| new_array << proc.call(val) unless proc.nil? }
+      my_each { |val| new_array << yield(val) if block_given? }
+      new_array
+    else
+      to_enum
+    end
+  end
```
Comment (mosesogwo) [inline myenumerables.rb]: You should have a single my_map method that can take an optional proc. You don't have to separate the methods.

File: myenumerables.rb [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+  def my_map_with_proc(proc = nil)
+    if block_given? || !proc.nil?
+      new_array = []
+      my_each { |val| new_array << proc.call(val) unless proc.nil? }
+      my_each { |val| new_array << yield(val) if block_given? }
+      new_array
+    else
+      to_enum
```
Comment (mosesogwo) [inline myenumerables.rb]: The logic here does not produce correct results. A more correct logic will be
```
Declare new array
If proc
  run my_each with proc.call
  return new array
elsif block_given?
  run my_each with yield
  return new_array
else
 to_enum
end
```

**Review by mosesogwo — CHANGES_REQUESTED** [review]
**Hi @chirchir12, good start on this project:+1:. Here are the recommended changes:**

:point_right: **my_all?** should be able to take an optional argument... the argument could be a Class, a Regex or any other pattern (a single character, word, etc). You can check this [doc](https://ruby-doc.org/core-2.6.5/Enumerable.html#method-i-all-3F) for the behaviour of the method in these scenarios.

:point_right:  When no block is given to **my_any?**, it should return `true` **only** if there is at least one element that is a truthy.
- `[false, nil, false].my_any?` should return `false`

:point_right: Just like in the case of **my_all?** above, Both **my_any?** and **my_none?** should be able to take an optional argument, which could be a Class, Regex or any other pattern.

:point_right: When no block or  argument is given, **my_none?** should return true if none of the members is true.
- `[false, nil, false].my_none?` should return `true`

:point_right: **my_inject** currently throws an error when called with a block on a range.
 - `(1..5).my_inject(4) { |prod, n| prod * n }` - should return 480 and not an error.


_**Kindly make these changes and submit another code review request.**_

**Any questions? You can reach me on Slack (Moses Ogwo)**

**Cheers:stars:**

### Status: Changes Requested.

**Review by rOluochKe — CHANGES_REQUESTED** [review]
Hello Team!
### You have done really good and for also making changes from the previous reviews so far :clap: Just one more method to go.
- [ ] my_map is not passing the test case
map
### Make the change and submit for another code review. good luck.
- Happy coding!

By Raymond
slack @iamraymond

**Comment by chirchir12** [comment]
### implemented changes as requested. Kindly check it on

**Review by rOluochKe — APPROVED** [review]
Hello Team!
### You have done well :clap: The project looks good, therefore, it has been approved. Congratulations :boom: :100:
### You can merge and submit a project completion form from the dashboard. Good luck with the next project. 
- Happy coding!

**Pull request merged**